lib/ProductOpener/Food.pm

```diff
@@ -2239,45 +2244,53 @@ sub compute_nova_group($) {
 				my $tagid = $';
 
 				if (has_tag($product_ref, $tagtype, $tagid)) {
-
-					if ($options{nova_groups_tags}{$tag} > $product_ref->{nova_group}) {
-
-						# only move group 1 product to group 3, not group 2
-						if (not (($product_ref->{nova_group} == 2) and ($options{nova_groups_tags}{$tag} == 3))) {
-							$product_ref->{nova_group_debug} .= " -- $tag : " . $options{nova_groups_tags}{$tag} ;
-							$product_ref->{nova_group} = $options{nova_groups_tags}{$tag};
-						}
-					}
+					push @matching_tags, [$tagtype, $tagid, $options{nova_groups_tags}{$tag}];
 				}
-
 			}
 		}
 	}
 
-	# Also loop through categories and ingredients to see if the taxonomy has associated minimum NOVA grades
-	# Categories need to be first, so that we can identify group 2 foods such as salt, sugar, fats etc.
+
 	# Group 2 foods should then not be moved to group 3
 	# (e.g. sugar contains the ingredient sugar, but it should stay group 2)
 	
-	foreach my $tag_type ("categories", "ingredients") {
+	foreach my $tagtype ("categories", "ingredients", "additives") {
 
-		if ((defined $product_ref->{$tag_type . "_tags"}) and (defined $properties{$tag_type})) {
+		if ((defined $product_ref->{$tagtype . "_tags"}) and (defined $properties{$tagtype})) {
 
-			foreach my $tag (@{$product_ref->{$tag_type . "_tags"}}) {
+			foreach my $tagid (@{$product_ref->{$tagtype . "_tags"}}) {
 
-				if ( (defined $properties{$tag_type}{$tag})
-					and (defined $properties{$tag_type}{$tag}{"nova:en"})
-					and ($properties{$tag_type}{$tag}{"nova:en"} > $product_ref->{nova_group})
-					# don't move group 2 to group 3
-					and not (($properties{$tag_type}{$tag}{"nova:en"} == 3) and ($product_ref->{nova_group} == 2))
-					) {
-					$product_ref->{nova_group_debug} .= " --- $tag_type : $tag : " . $properties{$tag_type}{$tag}{"nova:en"} ;
-					$product_ref->{nova_group} = $properties{$tag_type}{$tag}{"nova:en"};
+				my $nova_group = get_property($tagtype, $tagid, "nova:en");
+
+				if (defined $nova_group) {
+					push @matching_tags, [$tagtype, $tagid, $nova_group];
 				}
 			}
 		}
 	}
 
+	# Assign the NOVA group based on matching tags
+	# Group 2 foods should then not be moved to group 3
+	# (e.g. sugar contains the ingredient sugar, but it should stay group 2)
+
+	my %seen_markers = ();
+
+	foreach my $matching_tag_ref (@matching_tags) {
+		my ($tagtype, $tagid, $nova_group) = @$matching_tag_ref;
+		if (($nova_group >= $product_ref->{nova_group})
+			# don't move group 2 to group 3
+			and not (($nova_group == 3) and ($product_ref->{nova_group} == 2))
```

Comment:
I would like to have clarification on this, this works because of the order in which we process things right ? config option then categories then ingredients then additives, isn't it ?

Then it is worth mentioning I think. (it twist my mind every time).

Reply:
Maybe make it more robust, by adding a condition on tag_type ?

Reply:
I'm changing it to have separate lists of matching tags for each NOVA group, it will be more clear.